python/pyarrow/array.pxi

```diff
@@ -3573,17 +3602,19 @@ class FixedShapeTensorArray(ExtensionArray):
           ]
         ]
         """
-        if not obj.flags["C_CONTIGUOUS"]:
-            raise ValueError('The data in the numpy array need to be in a single, '
-                             'C-style contiguous segment.')
+
+        permutation = (-np.array(obj.strides)).argsort(kind='stable')
+        if permutation[0] != 0:
+            raise ValueError('First stride needs to be largest to ensure that '
+                             'individual tensor data is contiguous in memory.')
 
         arrow_type = from_numpy_dtype(obj.dtype)
-        shape = obj.shape[1:]
-        size = obj.size / obj.shape[0]
+        shape = np.take(obj.shape, permutation)
+        values = np.ravel(obj, order="K")
```

Comment:
Note that `ravel` can make a copy of the input if it is not contiguous, is this something that we want?

Reply:
Hmm, I wouldn't be so sure, example:
```python
>>> obj = np.arange(24).reshape(2,3,4)[:,::2]
>>> obj
array([[[ 0,  1,  2,  3],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [20, 21, 22, 23]]])
>>> obj.strides
(96, 64, 8)
>>> permutation = (-np.array(obj.strides)).argsort(kind='stable')
>>> permutation
array([0, 1, 2])   # permutation is ok
>>> values = np.ravel(obj, order="K")
>>> values
array([ 0,  1,  2,  3,  8,  9, 10, 11, 12, 13, 14, 15, 20, 21, 22, 23])
>>> values[0] = 999
>>> values
array([999,   1,   2,   3,   8,   9,  10,  11,  12,  13,  14,  15,  20,
        21,  22,  23])
>>> obj
array([[[ 0,  1,  2,  3],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [20, 21, 22, 23]]])

# values is a copy! which we can also check using:
>>> values.ctypes.data == obj.ctypes.data
False   # base addresses are different
```

Of course, we may not really care about this, since the conversion is probably correct anyway. But we may want to add a mention in the docstring that the conversion is only zero-copy if the input array is contiguous.

Reply:
Added a note to docstring: `If input array data is not contiguous a copy will be made.`.
Alternatively we could do `as_strided` and (hopefully) avoid copying:
```suggestion
        values = np.lib.stride_tricks.as_strided(
            obj, shape=(obj.size,), strides=(obj.dtype.itemsize,))
```

Reply:
`as_strided` can be a later PR if desired. The docstring addition is good for now!